### PR TITLE
Add all logic with ngrx for absence

### PR DESCRIPTION
**Pull request by f0xer1**
New deployed calendar view: https://6707de534b1ee13a5f7e91a9--stupendous-souffle-cd7e8d.netlify.app/

Added the ability to create absences by year with a limited number of them, added their display of two types on the calendar and sitebar. Absences can be added, updated and deleted.

In order to create absences for different years, you must be specifically on this year in the calendar.

Look lile: 
image
image
image
image
image

### CODE REVIEW

File: src/app/modules/calendar/calendar-month/calendar-month.component.ts [hunk 1]
```diff
@@ -51,4 +62,11 @@ export class CalendarMonthComponent implements OnInit {
     this.calendarService.setCurrentDate();
     this.calendar = this.calendarService.createCalendar();
   }
+
+  handleSickDayClick(day: CalendarItem) {
```
Comment (AlexanderAndreyev) [inline src/app/modules/calendar/calendar-month/calendar-month.component.ts]: we can handle only sick days? what about vacations?

File: src/app/modules/forms/absence-form/absence-form.component.html [hunk 2]
```diff
@@ -1,44 +1,68 @@
...
+                   [max]="currentDate.value.endOf('year').toISOString()"
+                   [matDatepicker]="pickerFrom" formControlName="dateFrom">
+            <mat-datepicker-toggle matSuffix [for]="pickerFrom"></mat-datepicker-toggle>
+            <mat-datepicker #pickerFrom startView="month"></mat-datepicker>
+          </mat-form-field>
+          <mat-form-field appearance="fill" class="absence-date-box-container">
+            <mat-label>To</mat-label>
+            <input matInput
+                   [min]="currentDate.value.startOf('year').toISOString()"
+                   [max]="currentDate.value.endOf('year').toISOString()"
```
Comment (AlexanderAndreyev) [inline src/app/modules/forms/absence-form/absence-form.component.html]: please create pipe for this manipulations, it's a bad practice to call a function inside the template

File: src/app/modules/forms/absence-form/absence-form.component.ts [hunk 3]
```diff
@@ -22,36 +40,108 @@ import {MatSelectModule} from "@angular/material/select";
...
+    NgIf
   ],
   templateUrl: './absence-form.component.html',
   styleUrls: ['./absence-form.component.css']
 })
 export class AbsenceFormComponent {
+  form: FormGroup;
+  isEditMode: boolean;
+  currentDate = new BehaviorSubject(moment());
+  private destroyRef = inject(DestroyRef);
```
Comment (AlexanderAndreyev) [inline src/app/modules/forms/absence-form/absence-form.component.ts]: if you use inject - please use it everywhere and avoid using injection in constructor. Or opposite :)

File: src/app/modules/sidebar/sidebar.component.html [hunk 4]
```diff
@@ -57,8 +52,7 @@ <h6 class="expansion-item-title">Break for {{ currentDate.value.year() }}</h6>
                 <span class="status-dot status-dot--red"></span> Sick
               </mat-panel-title>
               <mat-panel-description>
-                <!-- TODO: Replace 'X days left' with a dynamic value based on the days off logic -->
-                X days left
+                {{10 - counts.sickCount }}  days left
```
Comment (AlexanderAndreyev) [inline src/app/modules/sidebar/sidebar.component.html]: please move 10 to some const

File: src/app/store/absence.reducer.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+export const initialState: AbsenceState[] = [];
+
+const findOrCreateYearState = (state: AbsenceState[], year: number): AbsenceState => {
+  const yearState = state.find(absence => absence.year === year);
+  return yearState
+    ? yearState
+    : {year, absence: [], sickCount: 0, vacationCount: 0};
+};
+
+const updateCounts = (state: AbsenceState, days: number, absenceType: string, operation: 'add' | 'subtract') => {
```
Comment (AlexanderAndreyev) [inline src/app/store/absence.reducer.ts]: updateCounts -> getUpdatedCounts?

the name is a bit misleading - we don't update smth in this function, we just get a data

File: src/app/store/absence.reducer.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    ];
+  }),
+  on(updateAbsence, (state, {year, id, changes, daysBefore, daysAfter, from, to}) => {
+    const yearState = findOrCreateYearState(state, year);
+    let newSickCount = yearState.sickCount;
+    let newVacationCount = yearState.vacationCount;
+
+    if (from !== to) {
+      const sickAdjustment = from === 'sick' ? -daysBefore : daysAfter;
+      const vacationAdjustment = from === 'vacation' ? -daysBefore : daysAfter;
```
Comment (AlexanderAndreyev) [inline src/app/store/absence.reducer.ts]: 'sick' and 'vacation' should be enum :)

File: src/app/store/absence.selectors.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+export const selectFeature = (state: AppState) => state.absences;
+
+export const selectAllAbsenceByYear = (year: number) =>
+  createSelector(
+    selectFeature,
+    (state: AbsenceState[]) => {
+      return state.find(absence => absence.year === year)?.absence || []
+    }
+  );
+export const selectAll =
```
Comment (AlexanderAndreyev) [inline src/app/store/absence.selectors.ts]: selectAll is not really clear and generic. Can we make it a bit more specific?

**Review by AlexanderAndreyev — CHANGES_REQUESTED** [review]

**Comment by AlexanderAndreyev** [comment]
Also I noticed some bugs:
1) Can't delete created absence on the first delete action (it's still on the calendar)
2) I should be able to create an absence by clicking on the calendar cell
3) Update action does no work properly - I can update an absence but after modal window reopening notes section is not updated

File: src/app/store/absence.selectors.ts (hunk 7 above)
Comment (AlexanderAndreyev) [reply src/app/store/absence.selectors.ts]: I mean naming :)

File: src/app/modules/sidebar/sidebar.component.html [hunk 8]
```diff
@@ -52,26 +52,26 @@ <h6 class="expansion-item-title">Break for {{ currentDate.value.year() }}</h6>
...
               </mat-panel-title>
               <mat-panel-description>
-                {{10 - counts.sickCount }}  days left
+                {{ MAX_COUNT_OF_ABSENCE - counts.sickCount }}  days left
               </mat-panel-description>
             </mat-expansion-panel-header>
             <mat-list>
               <mat-list-item>
-                <h6 class="expansion-item-title">Break for {{ currentDate.value.year() }}</h6>
+                <h6 class="expansion-item-title">Break for {{ currentDate| dateFormat| async }}</h6>
```
Comment (AlexanderAndreyev) [inline src/app/modules/sidebar/sidebar.component.html]: is it observable? do you need async pipe here?

File: src/app/modules/calendar/calendar-month/calendar-month.component.ts [hunk 9]
```diff
@@ -34,39 +35,59 @@ export class CalendarMonthComponent implements OnInit {
...
-    this.calendar = this.calendarService.createCalendar();
+    this.calendar = this.calendarService.createCalendarForMonth();
   }
 
-  handleSickDayClick(day: CalendarItem) {
+  getSelectedMonth() {
+    this.calendar = this.calendarService.createCalendarForMonth();
+  }
+
+
```
Comment (AlexanderAndreyev) [inline src/app/modules/calendar/calendar-month/calendar-month.component.ts:71]: remove line please

File: src/app/modules/calendar/calendar-year/calendar-year.component.ts [hunk 10]
```diff
@@ -76,10 +77,26 @@ export class CalendarYearComponent implements OnInit {
...
   }
 
-  handleSickDayClick(day: CalendarItem) {
+  handleAbsenceDayClick(day: CalendarItem) {
+    const isUpdating = !!day.absence;
+
+    const absence: Absence = isUpdating
+      ? day.absence as Absence
+      : {
+        id: -1,
```
Comment (AlexanderAndreyev) [inline src/app/modules/calendar/calendar-year/calendar-year.component.ts:89]: -1 looks like magic number. Could you please add some explanation?

**Review by AlexanderAndreyev — CHANGES_REQUESTED** [review]

File: src/app/modules/calendar/calendar-year/calendar-year.component.ts (hunk 10 above)
Comment (f0xer1) [reply src/app/modules/calendar/calendar-year/calendar-year.component.ts:89]: If it's an edit of an absence, we send it, if you add a new one by clicking on the day, we send a fake absence to fill in the fields on the same day.

I'll add this as a comment, before I wrote a bigger code there that was easier to understand, but it was too big

**Review by AlexanderAndreyev — APPROVED** [review]

**Pull request merged**